### ISSUE TITLE
saving lists in hdf5 is slow

### ISSUE THREAD

**Issue by to266** [hyperspy/hyperspy#642]
as @francisco-dlp  wrote:
"..writing lists to hdf5 is currently slow and this is an issue because when saving files coming e.g. from the Osiris, the metadata always contain long lists that take several minutes to write even if the data itself is tiny"

**Comment by to266** [comment]
I've looked into this, and the problem is that HDF5 files do not have lists or tuples. 

As the file format supports (essentially) either arrays or dictionaries, at the moment we work around that by creating a dictionary with every element index as the key( `[0, 13., 'asd']` -> `{'0': 0, '1': 13, '3':'asd'}`)
The problem is then that creating dictionaries with so many key:value pairs is inefficient, when compared to saving (numpy) arrays.

The solution I've come up so far is to try to cast any lists or tuples as numpy arrays, and use the inefficient (but faithful) saving scheme only if the dtype is not 'Object'. If it is, instead save the result as an array (which can be done very efficiently and quickly).

The plus side is that it does not break old saved files and numpy-able structures are saved quickly and efficiently. 

The main drawback is that types are all homogenized if possible:

`[('tuple', 1), ('tuple', 2), ['list', 3]] ` -> `array([['tuple', '1'], ['tuple', '2'], ['list', '3']], dtype='|S5')`

but 

`[('tuple', 1), ('tuple', 2), ['list', 3, -30]] ` will stay the same (i.e. numbers are still numbers, tuples - tuples) due to numpy not being able to cast it as [3,2] array.

This makes things confusing. 
We could only use arrays when saving individual lists (i.e ndim=1), however as regular python lists can have mixed types and numpy (and hdf5) do not, there is no way to save those quickly and ensure that all objects are identical and of the original type..

__TLDR:__ Can either save lists and tuples faithfully and slow (what we do now), or "homogenize" them into the most general dtype in the list (i.e. all numbers get turned into strings if there is a string in the list) and save them quickly. 

Thoughts?

**Comment by to266** [comment]
Also, according to [this](http://docs.scipy.org/doc/numpy/reference/generated/numpy.can_cast.html#numpy.can_cast) if numpy always casts according to "safe" rules by default, even changing the datatype actually preserves full precision. So maybe not that terrible, just inconvenient..

**Comment by vidartf** [comment]
Do you have a file I can test with? I have an EDS map from the Osiris in Cambridge myself, but I'm not seeing any performance hit.

**Comment by to266** [comment]
Eh, I tried to reply to the email... Anyway, here's a snippet that shows the behaviour:

```python
s = hs.signals.Signal(range(10))
s.metadata.set_item('some_list', range(10000))
s.save('temp_file')
```

**Comment by francisco-dlp** [comment]
The figure below illustrates the problem and shows why what @to266 proposes should fix the problem.

![save_list_metadata](https://cloud.githubusercontent.com/assets/989719/9025160/79ea4be6-38ef-11e5-8d8e-1e5decff1eac.png)

**Comment by francisco-dlp** [comment]
Fixed in #646
